ERCS/erc-7540.md

```diff
@@ -319,19 +478,61 @@ The interface is fully backwards compatible with [ERC-4626](./eip-4626.md). The
 
 ## Reference Implementation
 
-WIP
+```solidity
+    // This code snippet is incomplete pseudocode used for example only and is no way intended to be used in production or guaranteed to be secure
+
+    mapping(address => uint256) public pendingDepositRequest;
+    
+    mapping(address => uint256) public claimableDepositRequest;
+
+    function requestDeposit(uint256 assets, address receiver, address owner, bytes calldata data) external returns (uint256 rid) {
+        require(assets != 0);
+        require(owner == msg.sender);
+
+        rid = 0; // no rid associated with this request
+
+        asset.safeTransferFrom(msg.sender, address(this), assets); // asset here is the Vault underlying asset
+
+        pendingDepositRequest[owner] += assets;
+
+        // Perform the callback
+        if (data.length != 0) {
+          ERC7540Receiver(receiver).onERC7540DepositReceived(msg.sender, owner, rid, data);
+        }
+
+        emit DepositRequest(msg.sender, operator, rid, assets);
+        return rid;
+    }
+
+    /**
+     * Include some arbitrary transition logic here from Pending to Claimable
+     */
+
+    function deposit(uint256 assets, address receiver) external returns (uint256 shares) {
+        require(assets != 0);
+
+        claimableDepositRequest[msg.sender] -= assets; // underflow would revert if not enough claimable assets
+
+        shares = convertToShares(assets); // this naive example uses the instantaneous exchange rate. It may be more common to use the rate locked in upon Claimable stage.
+
+        balanceOf[receiver] += shares;
+
+        emit Deposit(msg.sender, receiver, assets, shares);
+    }
+
+```
 
 ## Security Considerations
```

Comment:
Think this section is a bit choppy which could lead to some good info being glossed over. Suggest re-writing as follows:

In general, asynchronicity concerns make state transitions in the Vault much more complex and vulnerable to security risks. Access control on Vault operations, clear documentation of state transitioning, and invariant checks should all be performed to mitigate these risks. For example: 

- The view methods for viewing Pending and Claimable request states (e.g. `pendingDepositRequest`) are estimates useful for display purposes but can be outdated. The inability to know the final exchange rate will be on any Request requires users to trust the implementation of the asynchronous Vault in the computation of the exchange rate and fulfillment of their Request.

- Shares or assets locked for Requests can be stuck in the Pending state. Vaults may elect to allow for fungibility of pending claims or implement some cancellation functionality to protect users. 

Lastly, it is worth highlighting again here that the Claim functions for any asynchronous flows MUST enforce that `msg.sender == owner` to prevent theft of Claimable `assets` or `shares`.